src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -258,24 +239,266 @@ private IISAwaitable DoFlushAsync()
+                        }
+
+                        if (!buffer.IsEmpty)
+                        {
+                            // Write to IIS buffers
+                            // Guaranteed to write the entire buffer to IIS
+                            await WriteToIISAsync(buffer);
+                        }
+                        else if (readResult.IsCompleted)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            // Flush of zero bytes
+                            await FlushToIISAsync();
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        throw ex;
+                    }
+                    finally
+                    {
+                        // Always Advance the data pointer to the end of the buffer.
+                        Output.Reader.AdvanceTo(buffer.End);
+                    }
+                }
+
+               
+                // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                // two seperate loops
+                await CheckForUpgrade();
+
+                // Now we handle the read. 
+                var wb = Input.Writer.GetMemory();
+                _inputHandle = wb.Retain(true);
+
+                try
+                {
+                    // Lock around invoking ReadFromIISAsync as we don't want to call CancelIo
+                    // when calling read
+                    var read = await ReadFromIISAsync(wb.Length);
+
+                    // read value of 0 == done reading
+                    // read value of -1 == read cancelled, continue trying to read
+                    // but don't advance the pipe
+                    if (read == 0)
+                    {
+                        _doneReading = true;
+                    }
+                    else if (read != -1)
+                    {
+                        Input.Writer.Advance(read);
+                    }
+                }
+                finally
+                {
+                    // Always commit any changes to the Input pipe
+                    Input.Writer.Commit();
+                    _inputHandle.Dispose();
+                }
+
+                // Flush the read data for the Input Pipe writer
+                var flushResult = await Input.Writer.FlushAsync();
+
+                // If the pipe was closed, we are done reading, 
+                if (flushResult.IsCompleted || flushResult.IsCancelled)
+                {
+                    _doneReading = true;
+                }
+
+                // We are done reading now, set _reading to false.
+                // _reading may already be set to false through cancellation, but it is okay to be redundant here.
+                lock(_stateSync)
+                {
+                    _reading = false;
```

Comment:
Why don't you set this from inside of ReadFromIISAsync

Reply:
I would need to set it in OnAsyncCompletion instead (which is fine).